### ISSUE TITLE
[strict-boolean-expressions] Should not assume earlier conditions after await

### ISSUE THREAD

**Issue by Trinovantes** [typescript-eslint/typescript-eslint#3698]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```js
        '@typescript-eslint/strict-boolean-expressions': ['error', {
            allowNullableBoolean: true,
            allowNullableString: true,
        }],
```

```TS
interface State { field: string }
let state: State | null = { field: 'Hello world' }

function sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms))
}

function doSomethingWithState(state: State): void {
    console.log(state.field)
}

async function main(): Promise<void> {
    if (!state) {
        return
    }

    await sleep(1000) // Simulate a long computation task

    if (!state) { // Unexpected object value in conditional. The condition is always true.
        return
    }

    doSomethingWithState(state)
}

void main()
state = null
```

**Expected Result**

No lint error because it's not always true

**Actual Result**

Error

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.5` |
| `@typescript-eslint/parser`        | `4.28.5` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `14.17.2` |

**Comment by Trinovantes** [comment]
After looking into the issue a bit more, I think this is a Typescript issue as it doesn't seem to be discarding inferred type narrowing after `await`. The type of `state` in the second if-statement is *incorrectly* typed as `State` instead of `State | null`

https://github.com/microsoft/TypeScript/issues/27900

**Comment by bradzacher** [comment]
TS does not treat `await` or function call as signals to invalidate applied type narrowing.
This is an intentional limitation to improve devx.

Flow with its strict type system doesn't even do it for every case.
[Thought it will always do it for properties on objects.](https://flow.org/try/#0PQKgBAAgZgNg9gdzAZwC4CcCWBjVYTABQhmAdqgKbpQCG2FYAyqjZWAN5hSYUwAmALjAB+NFlIBzMAF9CfCthg10DKAFdSuTHFIoYFCgAcAFAFtkQ0mtMAjKgEohABXRxTmZBQA8ANziY+AD4AbkIaZABPTS4NLR0wUxoyYzRWCiFmNMcwFzcPbz8AwI5CMDBMKDAUlkoAOm5ePjAAXmawKxgYexKyspVUNXRSULLZUpQainqeflr9SVQAC2CwYGAwOABrGgjxmgQkvGR9I2MARgAGK-sRibTpxrmKBeXV9dhEMCpXdDBw9uUrgQZCki3+dmeYEUFGUFCaNgiegMJm62BonWIZQqVVSdQa-BabQ6XR6vX6g2G4zGZVxU3xfFqGEwpmMNzeG22uxpkwes3mEiWKzWXHgSG+cF+-1IgMQILAYOQYAhumhsPhiKZLNR6JghFkQA)
